tests/5.1/metadirective/test_metadirective_nothing.c

```diff
@@ -0,0 +1,98 @@
+//===--------------------- test_metadirective_nothing.c ---------------------===//
+//
+// OpenMP API Version 5.1 Nov 2020
+// 
+// Test for nothing directive within metadirectives. Runs a variety of
+// metadirectives that check if the nothing directive is properly rendered.
+// Primarily tests based on the fact that no matter what 'when' clause is 
+// rendered it should result in nothing, and thus no additional pragma should
+// be created. Thus, the threads should remain unchanged through this process
+// and the compiler should handle it properly. Handles both device and host
+// based tests.
+//
+////===---------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1024
+
+int metadirectiveOnDevice() {
+   int errors = 0;
+   int A[N];
+
+   for (int i = 0; i < N; i++) {
+      A[i] = 0;
+   }
+
+   #pragma omp target map(tofrom: A)
+   {
+      // We expect at least one of these when conditons to eval to true, thus having the nothing directive utilized
+      #pragma omp metadirective \
+         when( device={kind(nohost)}: nothing ) \
+         when( device={arch("nvptx")}: nothing) \
+         when( implementation={vendor(amd)}: nothing ) \
+         default( parallel for map(tofrom: A) )
+         {
+            for (int i = 0; i < N; i++) {
+               A[i] += omp_in_parallel();
+            }
+         }
+   }
+
+   for (int i = 0; i < N; i++) {
+      OMPVV_TEST_AND_SET(errors, A[i] != 0);
+   }
+
+   OMPVV_INFOMSG("Test ran with a number of available devices greater than 0");
+   OMPVV_INFOMSG_IF(A[0] == 0, "Test recognized device was of arch/vendor/kind nvidia, amd, or nohost");
+   OMPVV_WARNING_IF(A[0] == 1, "Test could not recognize if device was of arch/vendor/kind nvidia, amd or, nohost, even though there are devices available.");
+
+   return errors;
+}
+
+int metadirectiveOnHost() {
+  int errors = 0;
+  int A[N];
+
+  for (int i = 0; i < N; i++) {
+     A[i] = 0;
+  }
+
+  // We expect all of these when statements to eval to false, causing body of code to run in parallel
+  #pragma omp metadirective \
+     when( device={kind(host)}: nothing ) \
+     when( implementation={vendor(amd)}: nothing ) \
```

Comment:
don't use nothing here --  it needs to be a construct that won't be encountered...say something like:
when( construct={target}: distribute parallel for )

Since it is not called from the target directive, it should not execute.

Reply:
So we are trying to make this function match the functionality of the metadirectiveOnDevice() function.
The idea is, we **want** the when statements to evaluate to be true such that the code in the body of the function operates the same as OpenACC Serial.

We figured that by using the device={kind(host)} and implementation={vendor(amd)} that is it likely one of those will evaluate to true. Since there is not a lot of readily available information regarding context selectors, it is very tough to form when() statements that will evaluate to true when running on host.

We should either
1) Add more context-selectors to 'guarantee' that one of the when()'s in this function will evaluate to true and thus will operate like OpenACC serial (then we use the omp_is_parallel to tell whether the when()'s evaluate to true)
2) Revert to the previous commit **which does the opposite of the metadirectiveOnDevice function**. The opposite, meaning that the previous commit of this function expects the when()'s to evaluate to **false** and thus we expect the metadirective to run the default. This behavior doesn't really test 'nothing' properly, and we agreed in a few meetings ago we should shoot for option 1) to make the functions symmetrical and make the functions perform similar testing strategy

Reply:
I see what you are saying though ... it might be better for the metadirectiveOnHost function to have exactly what you said, directives that are not encountered! This way, we can make the default behavior **nothing** and then put a bunch of ridiculous context selectors that should not evaluate to true on the host (like arch=nvidia or device=nohost)

Reply:
This way we test both

- In the metadirectiveOnDevice, we test that **nothing** will work when paired with a **when** statement
- In the metadirectiveOnHost, we test that **nothing** will work when used as the **default** statement